core/src/main/java/org/kohsuke/stapler/framework/io/LargeText.java

```diff
@@ -310,7 +326,111 @@ public void doProgressText(StaplerRequest req, StaplerResponse rsp) throws IOExc
+        return accept.startsWith("multipart/form-data");
+    }
+
+    /**
+     * Add additional meta data to a streaming response.
+     * @param key The field to (over)write meta data for.
+     * @param value The meta data value.
+     */
+    protected void putStreamingMeta(String key, Object value) {
+        if (streamingMeta == null) streamingMeta = new JSONObject();
+        streamingMeta.put(key, value);
+    }
+
+    private long findNextLineStart(long start, long stop) throws IOException {
+        try (var f = source.open()) {
+            if (f.skip(start) != start) {
+                // The log file rolled over, send it in full.
+                return 0;
+            }
+            byte[] buf = new byte[64 * 1024];
+            long searchPosition = start;
+            int n;
+            while (searchPosition + 1 < stop && (n = f.read(buf)) > 0) {
+                for (int i = 0; i < n && searchPosition + 1 < stop; i++) {
+                    if (buf[i] == '\n') {
+                        // We have a match, send from after the \n.
+                        putStreamingMeta("startFromNewLine", true);
+                        return searchPosition + 1;
+                    }
+                    searchPosition++;
+                }
+            }
+        }
+        // fall back to original start.
+        return start;
+    }
+
+    private void doProgressTextStreaming(StaplerRequest2 req, StaplerResponse2 rsp) throws IOException {
+        setContentType(rsp);
+        String ct = rsp.getContentType();
+        if (ct == null || ct.isEmpty()) {
+            ct = "text/plain;charset=UTF-8";
+        }
+        if (ct.contains("\r") || ct.contains("\n")) {
+            throw new IOException("Found CR/LF in Content-Type. Aborting streaming mode");
+        }
+        String boundary = UUID.randomUUID().toString();
+        rsp.setContentType("multipart/form-data;boundary=" + boundary);
+        rsp.setStatus(HttpServletResponse.SC_OK);
+        putStreamingMeta("completed", completed);
+
+        try (var w = rsp.getWriter()) {
+            w.write("--" + boundary + "\r\n" // preamble for first part
+                    + "Content-Disposition: form-data;name=text\r\n"
+                    + "Content-Type: " + ct + "\r\n"
+                    + "\r\n");
+            long length = source.exists() ? source.length() : 0;
+
+            String s = req.getParameter("start");
+            long start = (s != null) ? Long.parseLong(s) : 0;
+            if (start > length) {
+                // text rolled over, send in full
+                start = 0;
+            } else if (start < 0 && length <= -start) {
+                // tail on small file, send in full
+                start = 0;
+            } else if (start < 0) {
+                // tail on large file, start at first new line in tail
+                start = findNextLineStart(length + start, length);
+            } else if (req.getParameter("searchNewLineUntil") != null) {
```

Comment:
I would make this a constant, 
```suggestion
            } else if (req.getParameter(SEARCH_STOP_PARAMETER) != null) {
```